StackOverflow in DispatchedContinuation

We are having a a bug in a class used to instantiate heavy objects in background thread. Seems to happen since bumping coroutines to 1.4.0. Looks like something is cycling in an endless loop.
A class:
```
class AsyncSingletonContainer<T>(scope: CoroutineScope, private val createObject: () -> T) {

    private val cachedSingletonObject = scope.async(Dispatchers.Default) { createObject() }

    suspend fun getInstance(): T = cachedSingletonObject.await()
}
```

Stack trace:
Fatal Exception: r4.a.i0: Fatal exception in coroutines machinery for DispatchedContinuation[Dispatchers.Default, Continuation at **.AsyncSingletonContainer$cachedSingletonObject$1.invokeSuspend(AsyncSingletonContainer.kt)@f734c38]. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
       at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.java:144)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.java:115)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.java:571)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.java:738)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.java:678)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.java:665)

Caused by java.lang.StackOverflowError: stack size 1037KB
       at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:452)
       at java.lang.StringBuilder.append(StringBuilder.java:137)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.DisposeOnCancel.toString(DisposeOnCancel.java:385)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CompletedContinuation.toString(CompletedContinuation.java:18)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.java:506)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(ResumeAwaitOnCompletion.java:1414)
       at java.lang.String.valueOf(String.java:2827)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlinx.coroutines.JobSupport.completeStateFinalization(JobSupport.java:320)
       at kotlinx.coroutines.JobSupport.tryFinalizeSimpleState(JobSupport.java:295)
       at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.java:853)
       at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.java:825)
       at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.java:111)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(BaseContinuationImpl.java:46)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.java:106)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.java:571)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.java:738)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.java:678)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.java:665)

I have a 100% reproducible case on a Samsung Galaxy S20+ running Android 10, with coroutines 1.4.1.  The stacktrace looks about the same as the others:

```
    kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[java.util.concurrent.Executors$FinalizableDelegatedExecutorService@312d609, Continuation at kotlinx.coroutines.channels.AbstractChannel.receiveOrClosed-ZYPwvRU(AbstractChannel.kt:624)@825027]){Closed(null)}@11232d4. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
        at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:144)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:115)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: java.lang.StackOverflowError: stack size 1039KB
   ...
        at kotlinx.coroutines.CancellableContinuationImpl.toString(CancellableContinuationImpl.kt:506)
        at java.lang.String.valueOf(String.java:2924)
E:     at java.lang.StringBuilder.append(StringBuilder.java:132)
        at kotlinx.coroutines.ResumeAwaitOnCompletion.toString(JobSupport.kt:1414)
        at java.lang.String.valueOf(String.java:2924)
        at java.lang.StringBuilder.append(StringBuilder.java:132)
        at kotlinx.coroutines.JobSupport.notifyCompletion(JobSupport.kt:1536)
        at kotlinx.coroutines.JobSupport.completeStateFinalization(JobSupport.kt:323)
        at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:240)
        at kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath(JobSupport.kt:903)
        at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:860)
        at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:825)
        at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:111)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        	... 3 more
```

I built our project using the `job-node-tostring` branch and the stack overflow no longer happens.

--- a/kotlinx-coroutines-core/common/src/JobSupport.kt
+++ b/kotlinx-coroutines-core/common/src/JobSupport.kt
@@ -1151,8 +1151,6 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
         override fun invoke(cause: Throwable?) {
             parent.continueCompleting(state, child, proposedUpdate)
         }
-        override fun toString(): String =
-            "ChildCompletion[$child, $proposedUpdate]"
     }
 
     private class AwaitContinuation<T>(
@@ -1350,6 +1348,7 @@ internal abstract class JobNode<out J : Job>(
     override val isActive: Boolean get() = true
     override val list: NodeList? get() = null
     override fun dispose() = (job as JobSupport).removeNode(this)
+    override fun toString() = "$classSimpleName@$hexAddress[job@${job.hexAddress}]"
 }
 
 internal class NodeList : LockFreeLinkedListHead(), Incomplete {
@@ -1384,7 +1383,6 @@ private class InvokeOnCompletion(
     private val handler: CompletionHandler
 ) : JobNode<Job>(job)  {
     override fun invoke(cause: Throwable?) = handler.invoke(cause)
-    override fun toString() = "InvokeOnCompletion[$classSimpleName@$hexAddress]"
 }
 
 private class ResumeOnCompletion(
@@ -1392,7 +1390,6 @@ private class ResumeOnCompletion(
     private val continuation: Continuation<Unit>
 ) : JobNode<Job>(job)  {
     override fun invoke(cause: Throwable?) = continuation.resume(Unit)
-    override fun toString() = "ResumeOnCompletion[$continuation]"
 }
 
 private class ResumeAwaitOnCompletion<T>(
@@ -1411,7 +1408,6 @@ private class ResumeAwaitOnCompletion<T>(
             continuation.resume(state.unboxState() as T)
         }
     }
-    override fun toString() = "ResumeAwaitOnCompletion[$continuation]"
 }
 
 internal class DisposeOnCompletion(
@@ -1419,7 +1415,6 @@ internal class DisposeOnCompletion(
     private val handle: DisposableHandle
 ) : JobNode<Job>(job) {
     override fun invoke(cause: Throwable?) = handle.dispose()
-    override fun toString(): String = "DisposeOnCompletion[$handle]"
 }
 
 private class SelectJoinOnCompletion<R>(
@@ -1431,7 +1426,6 @@ private class SelectJoinOnCompletion<R>(
         if (select.trySelect())
             block.startCoroutineCancellable(select.completion)
     }
-    override fun toString(): String = "SelectJoinOnCompletion[$select]"
 }
 
 private class SelectAwaitOnCompletion<T, R>(
@@ -1443,7 +1437,6 @@ private class SelectAwaitOnCompletion<T, R>(
         if (select.trySelect())
             job.selectAwaitCompletion(select, block)
     }
-    override fun toString(): String = "SelectAwaitOnCompletion[$select]"
 }
 
 // -------- invokeOnCancellation nodes
@@ -1463,7 +1456,6 @@ private class InvokeOnCancelling(
     override fun invoke(cause: Throwable?) {
         if (_invoked.compareAndSet(0, 1)) handler.invoke(cause)
     }
-    override fun toString() = "InvokeOnCancelling[$classSimpleName@$hexAddress]"
 }
 
 internal class ChildHandleNode(
@@ -1472,7 +1464,6 @@ internal class ChildHandleNode(
 ) : JobCancellingNode<JobSupport>(parent), ChildHandle {
     override fun invoke(cause: Throwable?) = childJob.parentCancelled(job)
     override fun childCancelled(cause: Throwable): Boolean = job.childCancelled(cause)
-    override fun toString(): String = "ChildHandle[$childJob]"
 }
 
 // Same as ChildHandleNode, but for cancellable continuation
@@ -1483,7 +1474,5 @@ internal class ChildContinuation(
     override fun invoke(cause: Throwable?) {
         child.parentCancelled(child.getContinuationCancellationCause(job))
     }
-    override fun toString(): String =
-        "ChildContinuation[$child]"
 }
 

--- a/kotlinx-coroutines-core/common/src/selects/Select.kt
+++ b/kotlinx-coroutines-core/common/src/selects/Select.kt
@@ -339,7 +339,6 @@ internal class SelectBuilderImpl<in R>(
             if (trySelect())
                 resumeSelectWithException(job.getCancellationException())
         }
-        override fun toString(): String = "SelectOnCancelling[${this@SelectBuilderImpl}]"
     }
 
     @PublishedApi

--- a/kotlinx-coroutines-core/jvm/src/Future.kt
+++ b/kotlinx-coroutines-core/jvm/src/Future.kt
@@ -41,7 +41,6 @@ private class CancelFutureOnCompletion(
         // interruption flag and it will cause spurious failures elsewhere
         future.cancel(false)
     }
-    override fun toString() = "CancelFutureOnCompletion[$future]"
 }
 
 private class CancelFutureOnCancel(private val future: Future<*>) : CancelHandler()  {

